Robustly starting a transaction is unergonomic

First, thanks for working on these bindings for pcsc.

In [our project](https://gitlab.com/hkos/openpgp-card), we're trying to robustly start a transaction.  This means having a loop that calls [`pcsc::Card::transaction`](https://docs.rs/pcsc/2.6.0/pcsc/struct.Card.html#method.transaction), and if that returns `pcsc::Error::ResetCard`, then calling [`pcsc::Card::reconnect`](https://docs.rs/pcsc/latest/pcsc/struct.Card.html#method.reconnect) and retrying.  Something like this:

```
loop {
   match card.transaction() {
    Ok(tx) => return tx,
    Err(pcsc::Error::ResetCard) => {
      card.reconnect()?;
    }
    e @ Err(_) => return e,
  }
}
```

Since this is a fair amount of code, we want to wrap the complexity in a function (actually, [we wrap a lot of functionality](https://gitlab.com/nwalfield/openpgp-card/-/blob/reborrow/pcsc/src/lib.rs#L129) and this is just a method).

When we do this in a [function (`TxClient::new(mut card: PcScCard)`)](https://gitlab.com/nwalfield/openpgp-card/-/blob/reborrow/pcsc/src/lib.rs#L34-119), the borrow checker complains:

```text
error[E0597]: `card` does not live long enough
   --> pcsc/src/lib.rs:47:17
    |
34  | impl<'b> TxClient<'b> {
    |      -- lifetime `'b` defined here
...
47  |         let c = card.card();
    |                 ^^^^ borrowed value does not live long enough
...
53  |                 Ok(mut tx) => {
    |                    ------ assignment requires that `card` is borrowed for `'b`
...
118 |     }
    |     - `card` dropped here while still borrowed

error[E0499]: cannot borrow `*c` as mutable more than once at a time
  --> pcsc/src/lib.rs:50:23
   |
34 | impl<'b> TxClient<'b> {
   |      -- lifetime `'b` defined here
...
50 |             let res = c.transaction();
   |                       ^ `*c` was mutably borrowed here in the previous iteration of the loop
...
53 |                 Ok(mut tx) => {
   |                    ------ assignment requires that `*c` is borrowed for `'b`

error[E0499]: cannot borrow `*c` as mutable more than once at a time
  --> pcsc/src/lib.rs:93:25
   |
34 | impl<'b> TxClient<'b> {
   |      -- lifetime `'b` defined here
...
50 |             let res = c.transaction();
   |                       - first mutable borrow occurs here
...
53 |                 Ok(mut tx) => {
   |                    ------ assignment requires that `*c` is borrowed for `'b`
...
93 |                         c.reconnect(
   |                         ^ second mutable borrow occurs here
...
```

Doing the same thing in a [macro](https://gitlab.com/nwalfield/openpgp-card/-/blob/0_2/pcsc/src/lib.rs#L21-105), however, works fine.

As far as we can tell, this has to do with reborrowing mutable references.  Basically, it is not possible to clone a mutable reference.  So rust moves a mutable reference into the called function (e.g., `pcsc::Card::transaction`).  When the result of that function is returned, its lifetime extends beyond the lifetime of the loop, obviously.  Thus, the second iteration of the loop is not able to use the mutable reference again.  See [this explanation](https://stackoverflow.com/questions/65474162/reborrowing-of-mutable-reference) for details.

I think a possible solution to the problem would be to have transaction return the mutable reference on error.  Thus, there could be a variant of `transaction` that looks like this:

```
fn transaction2(&mut self) -> Result<Transaction<'_>, (&mut Self, Error)>
```

This probably makes the implementation of `transaction2` a bit less elegant, but it should help us avoid the reborrowing issue.

What do you think?

--- a/pcsc/src/lib.rs
+++ b/pcsc/src/lib.rs
@@ -1173,6 +1173,31 @@ impl Card {
         }
     }
 
+    /// Like `Card::transaction`, but returns the reference to `self` on error.
+    ///
+    /// This function is like [`Card::transaction`], but also returns
+    /// the reference to `self` on error.  When starting a
+    /// transaction, it is necessary to deal with transient errors,
+    /// like [`Error::ResetCard`] by reconnecting to the card, and
+    /// retrying the transaction.  When this functionality is wrapped,
+    /// this doesn't work, because mutable references can't be
+    /// reborrowed.  This function returns the reference, which allows
+    /// this construct.
+    pub fn transaction2(
+        &mut self,
+    ) -> Result<Transaction, (&mut Self, Error)> {
+        // We can't use try_pcsc, because it returns an incompatible
+        // type.  And we can't wrap it in a closure due to lifetimes.
+        // Hence we inline try_pcsc!.
+
+        match unsafe { ffi::SCardBeginTransaction(self.handle) } {
+            ffi::SCARD_S_SUCCESS => Ok(Transaction {
+                card: self,
+            }),
+            err => Err((self, Error::from_raw(err))),
+        }
+    }
+
     /// Reconnect to the card.
     ///
     /// This function wraps `SCardReconnect` ([pcsclite][1], [MSDN][2]).

